### PR TITLE
Improve return value of `resolveConfig`, unwrap `ResolvableTo`

### LINKED ISSUE #9929: resolveConfig returns wrong TypeScript type for theme keys
**What version of Tailwind CSS are you using?**

3.2.4

**What build tool (or framework if it abstracts the build tool) are you using?**

Vite 3.2.4

**What version of Node.js are you using?**

16.18.1

**What browser are you using?**

N/A

**What operating system are you using?**

Linux

**Reproduction URL**

[Repro on CodeSandbox](https://codesandbox.io/s/wizardly-chandrasekhar-sk6ln7?file=/src/App.vue) (may take a second or click in the editor for the IDE to pick up the type issue)

**Describe your issue**

Using the [recently exposed resolveConfig types](https://github.com/tailwindlabs/tailwindcss/pull/8924), many config keys (such as `theme.colors`) throw TS errors:

Unresolved config key type

`ResolvableTo<T>` is an internal type that equals `T | ((utils: PluginUtils) => T)` - presumably a type that gets resolved during `resolveConfig`. The expected colour key (gray, in this case) doesn't exist on the function type, causing the issue. I suspect the resolved config should return the expected `RecursiveKeyValuePair` type with `ResolvableTo` removed.

### PR DESCRIPTION
**Pull request by RobinMalfait**
This PR improves the return type fo the `resolveConfig` a little bit.

In the config, a lot of values can be written in various forms (and most of it is optional). We also allow for dynamic "runtime" presets, so generating the perfect TypeScript types are almost impossible.

However, some of the parts in the config accept a function that resolves to type `T`, this improvement gets rid of the callback function and results in `T` directly (because calling `resolveConfig` should resolve this).

Fixes: #9929

### CODE REVIEW

File: resolveConfig.d.ts [hunk 1]
```diff
@@ -1,3 +1,12 @@
-import type { Config } from './types/config'
-declare function resolveConfig(config: Config): Config
+import type { Config, ResolvableTo } from './types/config'
+
+type UnwrapResolvables<T> = {
+  [K in keyof T]: T[K] extends ResolvableTo<infer R> ? R : T[K]
+}
+
+type ResolvedConfig<T extends Config> = Omit<T, 'theme'> & {
+  theme: UnwrapResolvables<T['theme']>
```
Comment (thecrypticace) [inline resolveConfig.d.ts:8]: You're gonna want to apply the same unwrapping to all the keys inside `theme.extend`

Comment (thecrypticace) [reply resolveConfig.d.ts:8]: Actually, let me re-word this. You need to unwrap all the keys inside theme.extend and then spread them into the top level object. Is that even possible??

**Review by thecrypticace — CHANGES_REQUESTED** [review]
General approach looks good sans the change for handling `theme.extend` unwrapping and merging.

File: resolveConfig.d.ts (hunk 1 above)
Comment (thecrypticace) [reply resolveConfig.d.ts:8]: Scratch this. It'll require some ridiculousness with conditional types and what not. Nevermind.

**Review by thecrypticace — APPROVED** [review]

**Comment by justinh-kairos** [comment]
I'm still seeing this with tailwind 3.3.2:
```
Property 'colors' does not exist on type 'UnwrapResolvables<{ container: { center: true; padding: string; screens: { "2xl": string; }; }; extend: { fontFamily: { sans: [string, ...string[]]; display: [string, ...string[]]; }; colors: { input: string; ring: { DEFAULT: string; offset: string; }; ... 10 more ...; card: { ...; }; }; borderRadius: { ...; }; keyf...'.ts(2339)
```

**Comment by stefusilviu** [comment]
I confirm the same on 3.4.1

image

**Comment by theocerutti** [comment]
No it still doesn't work:

```
const config = resolveConfig(tailwindConfig);

const mycolor = config.theme.colors.neutral['750']; // Element implicitly has an 'any' type because expression of type '750' can't be used to index type
```

**Pull request merged**